Fix grammar bug with XML namespaces that are also valid HTML tags and tag names that start with a valid HTML tag before a hyphen

MoritzKn opened the pull request:
### Description of the Change

Using XML namespaces in HTML is generally tolerated by the grammar even though it's not valid HTML. For this reason the `begin` regex in line 328 contains a colon. However when using a namespace that is also a valid HTML tag, this rule won't apply and instead a rule for a HTML tag specific scope like `meta.tag.inline.any.html` will apply and only capture the namespace as tag name and capture the actual tag name as attribute. This PR adds a lookbehind after the regexes for specific HTML tags to make sure only
actual tag names are captured and no namespaces.
This PR also adds specs for this behavior.

Before this PR, this:
```
<div:foo>
```
would be falsely tokenized as:
* `<`: `text.html.basic`, `meta.tag.block.any.html`, `punctuation.definition.tag.begin.html`
* `div`: `text.html.basic`, `meta.tag.block.any.html`, `entity.name.tag.block.any.html`
* `:foo`: `text.html.basic`, `meta.tag.block.any.html`, `entity.other.attribute-name.html`
* `>`: `text.html.basic`, `meta.tag.block.any.html`, `punctuation.definition.tag.end.html`

Now it will be tokenized as just:
* `<`: `text.html.basic`, `meta.tag.block.any.html`, `punctuation.definition.tag.begin.html`
* `div:foo`: `text.html.basic`, `meta.tag.other.html`, `entity.name.tag.other.html`
* `>`: `text.html.basic`, `meta.tag.block.any.html`, `punctuation.definition.tag.end.html`


### Alternate Designs
- None

### Benefits
- This fixes a problem with the [JSP Grammar](https://github.com/atom/language-java/blob/master/grammars/java%20server%20pages%20(jsp).cson) which is based on the HTML grammar. JSP allows custom tags which are written as `<ns:tag>` and if the namespace is for example `a`, this caused problems.

### Possible Drawbacks
- I am aware of none

### Applicable Issues
- I am aware of none

MoritzKn commented:
I just realized this problem also applies for hyphens and added another commit to fix that.
I also found a issue for this: #123

50Wliu commented:
If I am understanding the problem correctly, would the following change to the non-script/style rules produce the same result?
* Replacing `(?:[:-])\\b` with `(?=\\s|/?>)`

MoritzKn commented:
Yeah, should work too. Why would you prefer this?

50Wliu commented:
We have slowly been transitioning away from using `\\b` to denote "end-of-word-match".  It works well in text documents, but not as well in many programming languages.  In addition, the positive lookahead is clearer in its purpose: **only** allow a space or a closing bracket.  Contrast that with `\\b` (or the negative lookahead + `\\b`), which is harder to understand at first glance.

MoritzKn commented:
Makes sense. Should I change it?

50Wliu commented:
Not quite yet - let me figure out something for the script/style tags so that you only need to push one commit.

50Wliu commented:
Okay - I think it should be safe to remove the `(?![^>]*/>)` in those patterns.  See if you can find a problem with the following regexes:
* `(?:^\\s+)?(<)((?i:style))(?=\\s|>)`
* `(?:^\\s+)?(<)((?i:script))(?=\\s+[^>]*\\btype\\s*=\\s*[\'"]?text/(x-handlebars|(x-(handlebars-)?|ng-)?template|html)[\'"]?)`
* `(?:^\\s+)?(<)((?i:script))(?=\\s+[^>]*\\btype\\s*=\\s*[\'"]?text/coffeescript[\'"]?)`
* `(?:^\\s+)?(<)((?i:script))(?=\\s|>)`

MoritzKn commented:
@50Wliu I had to alter two of your regexes to prevent the pattern from matching self closing tags. I also added some tests for that. I'm not entirely happy with my solution but it's the best i came up with.

50Wliu commented:
I also thought about self-closing tags, but decided to use the simplified pattern because neither `style` nor `script` can be self-closing.

MoritzKn commented:
On the one hand the regex is pretty complex right now and a simpler one would be nice, on the other hand if someone still uses self closing `style` or `script` tags, it will break everything without this check.
So because I have no clear opinion, how about just leaving it how it is right now?

50Wliu commented:
@MoritzKn sorry for the delayed response.  I think the number of people who are using self-closing style and script tags is so low that we won't have to worry about them (in addition, why would you ever have `<style/>` or `<script/>`? What does that even do?).

MoritzKn commented:
@50Wliu:

> Sorry for the delayed response.

No problem. You're doing a great job managing all the community contributions. This sure is a whole lot of work; many thanks for that! 🥇 

> I think the number of people who are using self-closing style and script tags is so low that we won't have to worry about them

I updated the PR, but once I rebased I noticed that  6916c01 already fixed the issue for those tags, so I just left it like it was.

50Wliu commented on spec/html-spec.coffee:
```diff
@@ -414,3 +414,98 @@ describe 'HTML grammar', ->
+    it "doesn't tokenize XML namespaces as tags if the prefix is a valid style tag", ->
+      lines = grammar.tokenizeLines '<style:foo>'
+      expect(lines[0][1].value).toNotEqual 'style'
+      expect(lines[0][1].scopes).not.toContain 'entity.name.tag.style.html'
```
This is dangerous because this spec will continue to pass if the scope is changed to something else and the behavior regresses.  I think your `.toNotEqual` assertion is ok, but this one probably needs to be changed to a positive `.toEqual`.  Ditto for all the others.

MoritzKn commented:
Good point. I changed the spec.

50Wliu commented:
Thanks!
